returnn/training.py

```diff
@@ -177,10 +184,26 @@ def __init__(
             "time": time_rqmt,
         }
 
-        if self.use_horovod:
-            self.rqmt["cpu"] *= self.horovod_num_processes
-            self.rqmt["gpu"] *= self.horovod_num_processes
-            self.rqmt["mem"] *= self.horovod_num_processes
+        if self.pe_num_processes:
```

Comment:
There should be an assert that checks if Sisyphus is actually configured for the SGE, and not Slurm or local execution.

Reply:
I'm not sure we really have access to the engine at this point?

Maybe this check should be elsewhere? E.g. in the engine, assert that there are no `qsub` args?

Or maybe `rqmt` should be extended by an extra `pe` flag or so?

I'm a bit afraid that every solution will have drawbacks and not really be generic and later cause problems.

Reply:
well we might want to do multi-node multi-gpu training also at places where Slurm is used for job management.

so maybe 
* assert that with pe_num_processes > 1 not local engine is configured
* add handling for slurm (`rqmt["sbatch_args"]`) if slurm is configured.

(Ok, Slurm can be added in another PR)

Reply:
I have no idea how this works in Slurm. `sbatch_args` is specific to Slurm?

So in SGE, parallel execution means that the job is executed with the same rqmt on multiple nodes (but I think there are many possibilities on the exact scheme of that, or any constraints for it). Does Slurm have the same, or is the logic different there?

If it has the same, and this is maybe similar in most engines, we can add some `pe_num_procs` or so to `rqmt`, and then each engine should properly check and handle that.

To your specific suggestions: I think at this point we don't have access to the engine, or at least I'm not sure. Even if we have, I'm not sure we should make a job dependent on the engine, or whether the job should be generic and put such information into rqmt and then the engine should handle it, or throw an exception.

Reply:
Ok, but as I argued, I don't think checking this here and making the job engine-dependent is a good idea. So you disagree?

Reply:
But we can also add a helper function which has a more general coverage, like checking that `engine()` is actually returning an `EngineSelector` and otherwise directly testing for the return value, someting like:
```python
def engine_is_sge():
    engine = gs.engine()
    if instance(engine, EngineSelector):
        engine = gs.engine().get_used_engine_by_rqmt({})
    return isinstance(engine, SonOfGridEngine)
```

Reply:
> Ok, but as I argued, I don't think checking this here and making the job engine-dependent is a good idea. So you disagree?

I do agree with @albertz here, the job schould be general and backend agnostic.

Reply:
I'm not talking about making such check more generic. My argument is that the job should better be completely independent from the engine, not check anything about the engine. Rather, it should set some generic flag to `rqmt`, and then any engine should check later whether that flag can be handled and otherwise throw an exception.

Reply:
> So you disagree?

No, the code can also be in the `Task` itself. I am just saying we should not have this without checks.

Reply:
> My argument is that the job should better be completely independent from the engine

Totally agree as well. Then I am not sure what we are discussing here, it was your proposal to have it SGE specific. Then lets move this discussion over to Sisyphus on the engine independent "multi-node" parameter to add.

Reply:
So it seems like Slurm `sbatch` is somehow similar. Usually `mpirun` should know about the hosts automatically. SGE somehow tells `mpirun`, and I assume Slurm does the same.

I'm not so sure about the Slurm terminology. A Slurm node (`-N`) is a separate computer/hostname, as in SGE? SGE PE does not require that they are run on separate SGE nodes but they could also get scheduled on the same SGE node, wherever there are free resources. So Slurm "task" (`-n`) is actually like SGE PE?

So, how should we call the option in rqmt? `num_parallel_tasks`? And then the Slurm engine and SGE engine would handle that, and other engines would throw an exception.

Reply:
> it was your proposal to have it SGE specific.

No? I never said that. You argued for that ("There should be an assert that checks if Sisyphus is actually configured for the SGE"), I argued all the time against that, that we should not check or access the engine here.

Reply:
Okay, again from the beginning:

 - I am against having SGE specific code here at all. You added code and even a flag here that is SGE specific
 - I thought if you put SGE specific code here, there needs to be at least a check that SGE is actually used, this is all I argued.
 - Also yes, checking for sbatch_args and qsub_args in the Task itself makes sense.
 
But adding an independent rqmt flag (e.g. called just "nodes" or so) would make sense as well, instead of having SGE specific code here in the Training job.

Reply:
> even a flag here that is SGE specific

It's not really SGE specific. Or maybe only the name. But I actually did not know, I thought maybe it's called the same for other engines. I asked already, what other name do you suggest?

I'm not working on a Sis PR where the flag is called `parallel_tasks` in the `rqmt`. Probably the argument here in the job then also should be renamed accordingly.

> checking ... in the Task itself

What do you mean by that? Which code exactly? In the `Task` class? Why there? So far we are discussing here about either the job (`ReturnnTrainingJob`) or the engine (e.g. SGE `SonOfGridEngine`).

Reply:
Sis PR here: https://github.com/rwth-i6/sisyphus/pull/114

Reply:
Ok done.